### PR TITLE
[BugFix][Kernel] Fix Illegal memory access in causal_conv1d in H100

### PR DESCRIPTION
**Pull request by mzusman**
1. Fix illegal memory access in casusal_conv1d forward kernel, it was overlooked in #9189 since it's only apparent in H100 machines AFAIK .
2. Ease the diff tolerance constraint on bfloat16 test results in `test_selective_state_update_with_batch_indices` and match them to the constraints in `test_selective_state_update_with_heads_with_batch_indices` - Without this change, it also fails in H100 as apparently the diff is a little higher there.

CC @tlrmchlsmth

### CODE REVIEW

File: csrc/mamba/causal_conv1d/causal_conv1d.cu [hunk 1]
```diff
@@ -446,9 +446,12 @@ void causal_conv1d_fwd_kernel(ConvParamsBase params) {
         }
         else {
             // in case the final state is in between the threads data
-            reinterpret_cast<vec_t *>(x_vals_load)[1] = smem_exchange[last_thread + 1];
-            reinterpret_cast<vec_t *>(x_vals_load)[0] = smem_exchange[last_thread];
             const int offset = ((seqlen - (kWidth - 1)) % (kNElts));
+            if ((offset + kWidth - 2) >= kNElts){
+                // do not load to index 1 if we're not gonna read from there
+                reinterpret_cast<vec_t *>(x_vals_load)[1] = smem_exchange[last_thread + 1];
```
Comment (tlrmchlsmth) [inline csrc/mamba/causal_conv1d/causal_conv1d.cu:452]: Could you explain this a bit more?

**Review by tlrmchlsmth — COMMENTED** [review]
Thanks for the fix! 

Verified that `pytest tests/kernels/test_causal_conv1d.py` was hitting an illegal memory access, and that's now fixed on an H100

**Comment by tlrmchlsmth** [comment]
@mzusman, is `tests/kernels/test_causal_conv1d.py` passing for you on an H100? I am seeing a couple of torch.allclose failures and wondering if the tolerances need to be bumped up a bit

```
FAILED tests/kernels/test_causal_conv1d.py::test_causal_conv1d_varlen[False-64-2048-4-True-True-itype0] - AssertionError: assert False
FAILED tests/kernels/test_causal_conv1d.py::test_causal_conv1d_varlen[False-4096-2048-4-True-True-itype0] - AssertionError: assert False
```

**Review by tlrmchlsmth — APPROVED** [review]

**Comment by mzusman** [comment]
> @mzusman, is `tests/kernels/test_causal_conv1d.py` passing for you on an H100? I am seeing a couple of torch.allclose failures and wondering if the tolerances need to be bumped up a bit
> 
> ```
> FAILED tests/kernels/test_causal_conv1d.py::test_causal_conv1d_varlen[False-64-2048-4-True-True-itype0] - AssertionError: assert False
> FAILED tests/kernels/test_causal_conv1d.py::test_causal_conv1d_varlen[False-4096-2048-4-True-True-itype0] - AssertionError: assert False
> ```

I've dug into those 2 failures, and it seems to fail on a specific edge case where the final state data is split between the two iterations in the kernel (in the most of the cases the data ends up only on the last `smem_exchange` memory). The data in the `smem_exchange` gets overwritten in the last iteration so I've had to add another bulk of code to deal with this case inside the loop itself.

I guess it was overlooked previously because it fails on specific on sequence lengths that depend on the chunk size, while our tests generate sequence lengths randomly (for the varlen part, more specifically).
To catch this edge case, I've added sequence length = 1025 in `test_causal_conv1d` that fails without this fix.

File: csrc/mamba/causal_conv1d/causal_conv1d.cu (hunk 1 above)
Comment (mzusman) [reply csrc/mamba/causal_conv1d/causal_conv1d.cu]: Yeah, added more explanation

**Review by tlrmchlsmth — APPROVED** [review]
Tests are green for me now, thanks for the additional bug fix!
